src/posix/platform/netif.cpp

```diff
@@ -505,6 +510,182 @@ static void UpdateLink(otInstance *aInstance)
+    SuccessOrExit(AddRtAttr(&req.header, sizeof(req), RTA_DST, data, 16 /* IP6 address size in bytes */));
+    SuccessOrExit(AddRtAttr32(&req.header, sizeof(req), RTA_PRIORITY, kDefaultExternalRoutePriority));
+    SuccessOrExit(AddRtAttr32(&req.header, sizeof(req), RTA_OIF, ifIdx));
+
+    VerifyOrExit(send(sNetlinkFd, &req, sizeof(req), 0) > 0);
+
+    sAddedExternalRoutes[sAddedExternalRoutesNum++] = aPrefix;
+
+exit:
+    return;
+}
+
+static otError DeleteExternalRoute(const otIp6Prefix &aPrefix)
+{
+    struct
+    {
+        struct nlmsghdr header;
+        struct rtmsg    msg;
+        char            buf[512];
+    } req{};
+    unsigned char data[sizeof(in6_addr)];
+    char          addrBuf[OT_IP6_ADDRESS_STRING_SIZE];
+    unsigned int  ifIdx = otSysGetThreadNetifIndex();
+    otError       error = OT_ERROR_NONE;
+
+    VerifyOrExit(sNetlinkFd >= 0);
+
+    req.header.nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_NONREC;
+
+    req.header.nlmsg_len  = NLMSG_LENGTH(sizeof(rtmsg));
+    req.header.nlmsg_type = RTM_DELROUTE;
+    req.header.nlmsg_pid  = 0;
+    req.header.nlmsg_seq  = ++sNetlinkSequence;
+
+    req.msg.rtm_family   = AF_INET6;
+    req.msg.rtm_src_len  = 0;
+    req.msg.rtm_dst_len  = aPrefix.mLength;
+    req.msg.rtm_tos      = 0;
+    req.msg.rtm_scope    = RT_SCOPE_UNIVERSE;
+    req.msg.rtm_type     = RTN_UNICAST;
+    req.msg.rtm_table    = RT_TABLE_MAIN;
+    req.msg.rtm_protocol = RTPROT_BOOT;
+    req.msg.rtm_flags    = 0;
+
+    otIp6AddressToString(&aPrefix.mPrefix, addrBuf, OT_IP6_ADDRESS_STRING_SIZE);
+    inet_pton(AF_INET6, addrBuf, data);
+    SuccessOrExit(AddRtAttr(&req.header, sizeof(req), RTA_DST, data, 16 /* IP6 address size in bytes */));
+    SuccessOrExit(AddRtAttr32(&req.header, sizeof(req), RTA_OIF, ifIdx));
+
+    VerifyOrExit(send(sNetlinkFd, &req, sizeof(req), 0) > 0);
+
+exit:
+    return error;
+}
+
+static void UpdateExternalRoutes(otInstance *aInstance)
+{
+    for (int i = 0; i < static_cast<int>(sAddedExternalRoutesNum); ++i)
+    {
+        otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
+        otExternalRouteConfig config;
+        bool                  found = false;
+        while (otNetDataGetNextRoute(aInstance, &iterator, &config) == OT_ERROR_NONE)
+        {
+            if (otIp6IsAddressEqual(&config.mPrefix.mPrefix, &sAddedExternalRoutes[i].mPrefix) &&
+                config.mPrefix.mLength == sAddedExternalRoutes[i].mLength)
+            {
+                found = true;
+                break;
+            }
+        }
+        if (!found)
+        {
+            if (DeleteExternalRoute(sAddedExternalRoutes[i]) == OT_ERROR_NONE)
+            {
+                otIp6Prefix &x = sAddedExternalRoutes[i], &y = sAddedExternalRoutes[sAddedExternalRoutesNum - 1];
+                otIp6Prefix  tmp = x;
+                x                = y;
+                y                = tmp;
+                --sAddedExternalRoutesNum;
+                --i;
```

Comment:
We don't need to swap the ith element with the last one - simply overwrite the ith element with the last one and decrease `sAddedExternalRoutesNum` should work.